cluster-autoscaler/simulator/clustersnapshot/basic.go

```diff
@@ -241,16 +232,32 @@ func (snapshot *BasicClusterSnapshot) AddNodeInfo(nodeInfo *framework.NodeInfo)
 	return nil
 }
 
+// Initialize initializes the snapshot.
+func (snapshot *BasicClusterSnapshot) Initialize(nodes []*apiv1.Node, scheduledPods []*apiv1.Pod) error {
+	snapshot.Clear()
+
+	knownNodes := make(map[string]bool)
+	for _, node := range nodes {
+		if err := snapshot.AddNode(node); err != nil {
```

Comment:
The only error condition for adding a node is if your `[]*apiv1.Node` set has a duplicate. I wonder if there's a more efficient way of doing that targeted error handling earlier in the execution flow so we don't have to do so much error handling at this point. It would also have the side-benefit of allowing us to ditch this `knownNodes` accounting overhead in this function.

Reply:
IMO this is the perfect place to validate this:
* The alternative is for `Initialize()` to assume that some validation happened earlier and that its input is correct. This doesn't seem safe, as it relies on every `Initialize()` user properly validating data first.
* There are multiple places that call `Initialize()`, so ideally we'd want to extract the validation logic to remove redundancy anyway. If we're extracting it outside of `Initialize()`, we essentially have 2 functions that always need to be called in sequence.

Keep in mind that this should be called once per snapshot per loop, so the `knownNodes` overhead should be trivial compared to the rest of the loop.

Reply:
O.K. I think my initial skepticism here is keeping the source of truth in two places here:

1) in the cluster snapshot data store
2) in the local function (`knownNodes`)

I don't have any reason to suspect a race condition risk here, but would this be more "correct"?:

```golang
	for _, pod := range scheduledPods {
		if err := snapshot.getInternalData().addPod(pod, pod.Spec.NodeName); err != nil {
			continue
		}
	}
```

The only error return condition for the `addPod()` method is

```golang
	if _, found := data.nodeInfoMap[nodeName]; !found {
		return ErrNodeNotFound
	}
```

So rather than keeping a local accounting of "found nodes" (`knownNodes`) and conditionally calling `addPod()`, we could simply leverage the lookup in `addPod()` which is equivalently expensive to `if knownNodes[pod.Spec.NodeName]`, and by so doing permit ourselves to drop the `knownNodes` local accounting.